### PR TITLE
Add monitoring ADR

### LINKED ISSUE #8: ADR required for monitoring setup
We have made some progress with setting up monitoring, and have some issues already created in the appropriate repos for creating service monitors, prometheus / grafana deployments etc. What's lacking is a document that adds context to our setup and future plans. For this we should prepare an ADR that outlines our monitoring architecture.

### PR DESCRIPTION
**Pull request by 4n4nd**
we can discuss the monitoring solution here

fixes #8

### CODE REVIEW

File: docs/adr/0006-monitoring-structure.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+## Decision Drivers <!-- optional -->
+
+* Access to cluster-wide permissions
+  * The Prometheus instance might not have cluster wide resource access
+* Prometheus resource locations
+  * Where do we keep the monitoring resources like servicemonitors/podmonitors?
+  * Whose responsibility is to create these monitoring resources?
+* Complexity of Prometheus deployments
+  * How many instances of Prometheus do we deploy?
+  * Which namespaces do these Prometheus instances are deployed in?
```
Comment (HumairAK) [inline docs/adr/0006-monitoring-structure.md]: change to: 

> Which namespaces should these Prometheus instances be deployed in?

File: docs/adr/0006-monitoring-structure.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+How do we structure the Prometheus-Operator/Prometheus Deployments in
+Operate First?
+
+## Decision Drivers <!-- optional -->
+
+* Access to cluster-wide permissions
+  * The Prometheus instance might not have cluster wide resource access
+* Prometheus resource locations
+  * Where do we keep the monitoring resources like servicemonitors/podmonitors?
+  * Whose responsibility is to create these monitoring resources?
```
Comment (HumairAK) [inline docs/adr/0006-monitoring-structure.md]: > Whose responsibility is to

change to:
 
> Whose responsibility is it to

**Comment by HumairAK** [comment]
So this covers the prometheus aspect of monitoring, what about grafana/alerting?

**Comment by 4n4nd** [comment]
> So this covers the prometheus aspect of monitoring, what about grafana/alerting?

I was thinking we can create separate ADRs for visualization and alerting

File: docs/adr/0006-monitoring-structure.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
+# Application Monitoring in Operate First using Prometheus
+
+* Status: accepted
+* Deciders: 4n4nd, anishasthana, hemajv, HumairAK, tumido
+* Date: 2021-01-06
+
+Technical Story: [issue-1](https://github.com/operate-first/blueprint/issues/8)
+                 [issue-2](https://github.com/operate-first/apps/issues/57)
+
+## Context and Problem Statement
```
Comment (HumairAK) [inline docs/adr/0006-monitoring-structure.md:10]: I think we could use more information here to better understand the rest of the doc. For example I think some questions that might be worth answering:

- Why are we considering giving service accounts cross name permissions? Why aren't we using operatorgroups to do this? 
- Are we using ODH monitoring? 
- Are we using ODH to monitor non ODH services? 
- Why are we weighing options between multiple prometheus instances vs one

Basically, when I read the `Considered Options` below, I'm not sure why one ought to pick Option 1 over Option 2, because I don't know what our concerns/priorities are, etc.

File: docs/adr/0006-monitoring-structure.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+Since there will be multiple services deployed in the Operate First environment
+(ex. Jupyterhub, Argo, Superset, etc), we need to be able to monitor them. <br>
```
Comment (anishasthana) [inline docs/adr/0006-monitoring-structure.md]: We're not calling out ODH vs non-ODH here

File: docs/adr/0006-monitoring-structure.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+To do so we will use [Prometheus](https://prometheus.io/).
+
+How do we structure the Prometheus-Operator/Prometheus Deployments in
+Operate First?
+
+## Decision Drivers <!-- optional -->
+
+* Access to cluster-wide permissions
```
Comment (anishasthana) [inline docs/adr/0006-monitoring-structure.md:39]: What about api conflicts between cluster monitoring/user workload monitoring and the prometheus community operator?

File: docs/adr/0006-monitoring-structure.md (hunk 4 above)
Comment (4n4nd) [reply docs/adr/0006-monitoring-structure.md]: We should be able to monitor them all right? ODH services as well non-ODH services like Observatorium

Comment (anishasthana) [reply docs/adr/0006-monitoring-structure.md]: All of the services in this list are ODH, so someone _might_ think it's just caring about ODH.

Comment (4n4nd) [reply docs/adr/0006-monitoring-structure.md]: I'll add a non-ODH example

File: docs/adr/0006-monitoring-structure.md (hunk 3 above)
Comment (4n4nd) [reply docs/adr/0006-monitoring-structure.md:10]: added some more context

File: docs/adr/0006-monitoring-structure.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+(ex. Jupyterhub, Argo, Superset, Observatorium, etc) distributed over various
+namespaces, we need to be able to monitor them.
+
+To do so we will use [Prometheus](https://prometheus.io/).
+This will be deployed using the [ODH Operator](https://opendatahub.io/) (ODH
+operator includes the Prometheus Operator to do this).
+
+Normally, Prometheus Operator would have been able to monitor services across
+namespaces using cluster-wide access, but the recent versions of Openshift have
+an instance of Prometheus Operator pre-installed for cluster monitoring so we
```
Comment (HumairAK) [inline docs/adr/0006-monitoring-structure.md:22]: > but the recent versions of Openshift have
an instance of Prometheus Operator

change to: 

> but the recent versions of Openshift **has** an instance of Prometheus Operator

Comment (HumairAK) [reply docs/adr/0006-monitoring-structure.md:22]: > Normally, Prometheus Operator would have been able to monitor services across namespaces using cluster-wide access, but the recent versions of Openshift have an instance of Prometheus Operator pre-installed for cluster monitoring so we cannot do a cluster-wide Prometheus Operator setup since this would conflict with the cluster monitoring instance.

This is sentence is a bit long, can we break it up, suggestion: 

> Normally, Prometheus Operator would have been able to monitor services across namespaces using cluster-wide access, but the recent versions of Openshift have an instance of Prometheus Operator pre-installed for cluster monitoring. As such, we cannot do a cluster-wide Prometheus Operator setup since this would conflict with the cluster monitoring instance.

File: docs/adr/0006-monitoring-structure.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+cannot do a cluster-wide Prometheus Operator setup since this would conflict
+with the cluster monitoring instance.
+
```
Comment (HumairAK) [inline docs/adr/0006-monitoring-structure.md:25]: Sentence is runon change: 
> Normally, Prometheus Operator would have been able to monitor services across
namespaces using cluster-wide access, but the recent versions of Openshift have
an instance of Prometheus Operator pre-installed for cluster monitoring so we
cannot do a cluster-wide Prometheus Operator setup since this would conflict
with the cluster monitoring instance.

To something like: 
> Normally, Prometheus Operator would have been able to monitor services across
namespaces using cluster-wide access, but the recent versions of Openshift have
an instance of Prometheus Operator pre-installed for cluster monitoring. As such we
cannot do a cluster-wide Prometheus Operator setup since this would conflict
with the cluster monitoring instance.

Comment (4n4nd) [reply docs/adr/0006-monitoring-structure.md:25]: updated

File: docs/adr/0006-monitoring-structure.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+Since there will be multiple services deployed in the Operate First environment
+(ex. Jupyterhub, Argo, Superset, Observatorium, etc) distributed over various
+namespaces, we need to be able to monitor them.
+
+To do so we will use [Prometheus](https://prometheus.io/).
+This will be deployed using the [ODH Operator](https://opendatahub.io/) (ODH
+operator includes the Prometheus Operator to do this).
+
+Normally, Prometheus Operator would have been able to monitor services across
```
Comment (anishasthana) [inline docs/adr/0006-monitoring-structure.md:20]: I feel like we should also mention User Workload Monitoring

Comment (4n4nd) [reply docs/adr/0006-monitoring-structure.md:20]: I think since cluster-monitoring and UWM both use the same instance of Proemtheus-Operator, it is okay to to not include it here as the conflict is with a second instance of the operator and not UWM.

**Review by hemajv — APPROVED** [review]
This looks great! :ship: :+1:

File: docs/adr/0006-monitoring-structure.md (hunk 8 above)
Comment (anishasthana) [reply docs/adr/0006-monitoring-structure.md:20]: I'm commenting more from a "as of now UWM is not a common thing, but it will be in the near future"

Comment (4n4nd) [reply docs/adr/0006-monitoring-structure.md:20]: added something for it

File: docs/adr/0006-monitoring-structure.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+namespaces using cluster-wide access, but the recent versions of Openshift have
+an instance of Prometheus Operator pre-installed for cluster monitoring. As such we
+cannot do a cluster-wide Prometheus Operator setup since this would conflict
+with the cluster monitoring instance.
+
+We also cannot use Operator Groups to do a multiple namespace installation,
+since selection of multiple target namespaces using operator groups might be
+[deprecated in a future release](https://docs.openshift.com/container-platform/4.6/operators/understanding/olm/olm-understanding-operatorgroups.html#olm-operatorgroups-target-namespace_olm-understanding-operatorgroups).
+
+One possible solution is User Workload Monitoring, but it is still a
```
Comment (4n4nd) [inline docs/adr/0006-monitoring-structure.md:30]: @anishasthana Is this okay for UWM?

**Review by anishasthana — APPROVED** [review]
